### DISCUSSION TITLE
suggestion: todo implementing simple SharedPreferences storage

### DISCUSSION THREAD

**Discussion by lorriman** [rrousselGit/riverpod#1588] [Ideas]
It didn't seem immediately obvious how to create a provider for SharedPreferences. FutureProvider for example, seems like overkill and causes boilerplate in the UI code when yet the fetching of the SharedPreferences.instance is a one-time affair and nearly instant.

This could be included  as a simple example for the todo example app, thereby answering that and similar questions for the future.

**Comment by scopendo** [comment]
Just answered this at the bottom of https://github.com/rrousselGit/riverpod/discussions/1597#discussioncomment-3538655.

```dart
Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  final prefs = await SharedPreferences.getInstance();

  runApp(
     ProviderScope(
      overrides: [
        sharedPreferencesProvider.overrideWithValue(prefs),
      ],
      child: const MyApp(),
    ),
  );
}
```

I use a function `MustOverrideProvider` that serves as a faux type to simplify the definition of providers whose values must be overridden.

```dart
final sharedPreferencesProvider = MustOverrideProvider<SharedPreferences>();

/// ignore: non_constant_identifier_names
Provider<T> MustOverrideProvider<T>() {
  return Provider<T>(
    (_) => throw ProviderNotOverriddenException(),
  );
}

class ProviderNotOverriddenException implements Exception {
  @override
  String toString() {
    return 'The value for this provider must be set by an override on ProviderScope';
  }
}
```
